crypto/property/property.c

```diff
@@ -304,7 +304,11 @@ int ossl_method_store_add(OSSL_METHOD_STORE *store, const OSSL_PROVIDER *prov,
         impl->properties = ossl_parse_property(store->ctx, properties);
         if (impl->properties == NULL)
             goto err;
-        ossl_prop_defn_set(store->ctx, properties, impl->properties);
+        if (!ossl_prop_defn_set(store->ctx, properties, impl->properties)) {
+            ossl_property_free(impl->properties);
+            impl->properties = NULL;
+            goto err;
+        }
```

Comment:
I do have a concern here.
On line 330 below, `impl->properties` is referenced, even though that is now part of the `store->ctx`.  This may be fine, but it's a bit unusual.
A bigger concern is on lines 330- 338. There, `impl` is pushed onto the `alg->impls` stack, presumably with the properties still in tow, which are owned by `store->ctx`; while freeing the `alg->impls` stack does not free the properties (i.e. `impl_free()` does not free `properties`) is there a chance that the pointer to the properties become stale when `store` is freed?

Reply:
The property definition cache lives for as long as the libctx and the method store is always associated with that libctx. The property definition cache never flushes itself - once added entries stay in the cache for the duration of the cache lifetime (which is tied to the duration of the libctx lifetime). So the properties pointer should never become stale.

That said I do see this bit of code in `ossl_prop_defn_set` that looks a bit suspect:

https://github.com/openssl/openssl/blob/163bf682fd93971d07e66e3da339c229b86dc849/crypto/property/defn_cache.c#L111-L115

So, if when we add a new property definition there is already one for the same property definition then the *old* one gets freed and we use the new one instead. Therefore anything using the old reference will suddenly have a stale reference. In practice however this is a "should not happen" scenario because the only caller of `ossl_prop_defn_set` calls `ossl_prop_defn_get` immediately before hand. This is a different pre-existing issue to what is being addressed in this PR.

Reply:
Should there be an assert there?

Reply:
> Should there be an assert there?

Yeah - seems reasonable. I've added a new commit with an assert at that spot.